### PR TITLE
perf(element): add one sweep element creation

### PR DESCRIPTION
**Pull request by j-mendez**
1. major performance improvements on element creation

### CODE REVIEW

File: src/node.rs [hunk 1]
```diff
@@ -238,25 +237,24 @@ pub struct Element {
...
-                    .map(LocalName::from)
-                    .collect()
-            });
+    pub fn new(name: QualName, attributes: Vec<Attribute>) -> Self {
+        let mut classes: HashSet<LocalName> = HashSet::new();
+        let mut id: Option<LocalName> = Option::default();
+        let mut attrs = Attributes::with_capacity(attributes.len());
+
+        for a in attributes {
+            if a.name.local.deref() == "id" {
```
Comment (adamreichold) [inline src/node.rs:246]: Maybe `match a.name.local.deref() { .. }` to avoid the repetition?

File: src/node.rs [hunk 2]
```diff
@@ -238,25 +237,24 @@ pub struct Element {
...
-            .map_or_else(HashSet::new, |a| {
-                a.value
-                    .deref()
-                    .split_whitespace()
-                    .map(LocalName::from)
-                    .collect()
-            });
+    pub fn new(name: QualName, attributes: Vec<Attribute>) -> Self {
+        let mut classes: HashSet<LocalName> = HashSet::new();
+        let mut id: Option<LocalName> = Option::default();
```
Comment (adamreichold) [inline src/node.rs:242]: I think `None` would be less surprising when reading this?

Comment (j-mendez) [reply src/node.rs]: updated

File: src/node.rs (hunk 1 above)
Comment (j-mendez) [reply src/node.rs]: @adamreichold before I saw this, actually updated it to deref once to avoid the penalties.

File: src/node.rs [hunk 3]
```diff
@@ -238,25 +237,31 @@ pub struct Element {
...
+    pub fn new(name: QualName, attributes: Vec<Attribute>) -> Self {
+        let mut classes: HashSet<LocalName> = HashSet::new();
+        let mut attrs = Attributes::with_capacity(attributes.len());
+        let mut id: Option<LocalName> = None;
+
+        for a in attributes {
+            let name_local = a.name.local.deref();
+            let mut value = "";
+
+            if name_local == "id" {
```
Comment (cfvescovo) [inline src/node.rs:249]: As @adamreichold suggested, maybe use a `match` statement here

Comment (j-mendez) [reply src/node.rs:249]: updating now

File: src/node.rs (hunk 1 above)
Comment (adamreichold) [reply src/node.rs]: The `.deref()` call is basically free, from my PoV this is about readability. Hence, I also find the `value` dance too much as it obfuscates the code.

Comment (adamreichold) [reply src/node.rs]: > Hence, I also find the value dance too much as it obfuscates the code.

Ok, that seems to be gone anyway. Sorry for the noise.

**Comment by j-mendez** [comment]
@cfvescovo re-verted `a.value.deref()` back to prior. Can take a look later on updating to a match later or etc.

File: src/node.rs (hunk 1 above)
Comment (j-mendez) [reply src/node.rs]: reverted for readability since its minor for this

**Comment by cfvescovo** [comment]
Maybe we could `match a.name.local.deref()` and check if it is equal to `id` or `class`. The call to `a.value().deref()` should be inside each arm of our `match`. Performing it inside the `match` is better since it will only be called if needed (at most once per iteration).

**Comment by j-mendez** [comment]
> Maybe we could `match a.name.local.deref()` and check if it is equal to `id` or `class`. The call to `a.value().deref()` should be inside each arm of our `match`. Performing it inside the `match` is better since it will only be called if needed (at most once per iteration).

I like that, feel free to hop on the PR and make those changes or any, need to brb.

File: src/node.rs [hunk 4]
```diff
@@ -244,17 +244,11 @@ impl Element {
...
-                value = a.value.deref();
-                id = Some(LocalName::from(value));
-            }
-            if name_local == "class" {
-                if value.is_empty() {
-                    value = a.value.deref();
-                }
-                classes.extend(value.split_whitespace().map(LocalName::from));
+                id = Some(LocalName::from(a.value.deref()));
+            } else if name_local == "class" {
```
Comment (j-mendez) [inline src/node.rs]: was rushing and made this change without thinking. @adamreichold @cfvescovo  feel free to squash the commits for this PR.

Comment (cfvescovo) [reply src/node.rs]: Will do, thanks

**Comment by cfvescovo** [comment]
> > Maybe we could `match a.name.local.deref()` and check if it is equal to `id` or `class`. The call to `a.value().deref()` should be inside each arm of our `match`. Performing it inside the `match` is better since it will only be called if needed (at most once per iteration).
> 
> I like that, feel free to hop on the PR and make those changes or any, need to brb.

I do not think I have permissions to edit this PR

**Comment by j-mendez** [comment]
> match a.name.local.deref()

ok, ill make the changes really quick should be simple

**Comment by j-mendez** [comment]
> > > Maybe we could `match a.name.local.deref()` and check if it is equal to `id` or `class`. The call to `a.value().deref()` should be inside each arm of our `match`. Performing it inside the `match` is better since it will only be called if needed (at most once per iteration).
> > 
> > 
> > I like that, feel free to hop on the PR and make those changes or any, need to brb.
> 
> I do not think I have permissions to edit this PR

@cfvescovo updated

**Comment by cfvescovo** [comment]
@j-mendez thank you

**Pull request merged**